actionpack/lib/action_dispatch/routing/inspector.rb

```diff
@@ -33,11 +33,11 @@ def reqs
       end
 
       def controller
-        requirements[:controller] || ':controller'
+        parts.include?(:controller) ? ':controller' : requirements[:controller]
       end
 
       def action
-        requirements[:action] || ':action'
+        parts.include?(:action) ? ':action' : requirements[:action]
```

Comment:
The number of `parts` is almost always so low that the speed difference doesn't matter – right?